Source/Core/VideoCommon/VertexShaderGen.cpp

```diff
@@ -399,18 +399,33 @@ ShaderCode GenerateVertexShaderCode(APIType api_type, const vertex_shader_uid_da
       out.Write("o.colors_1 = color1;\n");
   }
 
-  // write the true depth value, if the game uses depth textures pixel shaders will override with
-  // the correct values
-  // if not early z culling will improve speed
-  if (g_ActiveConfig.backend_info.bSupportsClipControl)
+  if (g_ActiveConfig.backend_info.bSupportsDepthClamp)
+  {
+    // Since we're adjusting z for the depth range before the perspective divide, we have to do our
+    // own clipping. We want to clip so that -w <= z <= 0.
+    out.Write("o.clipDist0 = o.pos.z + o.pos.w;\n");  // Near: z < -w
+    out.Write("o.clipDist1 = -o.pos.z;\n");  // Far: z > 0
+
+    // We have to handle the depth range in the vertex shader, because some games will use a depth
+    // range beyond the normal depth range of 0..1.
+    out.Write("o.pos.z = o.pos.w * " I_PIXELCENTERCORRECTION".w - "
+              "o.pos.z * " I_PIXELCENTERCORRECTION".z;\n");
+  }
+  else
   {
+    // User-defined clipping is not supported, thus we rely on the API to handle the depth range.
```

Comment:
`gl_ClipDistance` is supported since OpenGL 1.3, so we can assume it is supported if it can run Dolphin. However to actually be able to make use of that feature for depth clipping we need to disable the incorrect depth clipping planes through `GL_DEPTH_CLAMP` which was introduced in OpenGL 3.2.

OpenGL ES on the other hand only supports gl_ClipDistance through a recent extension, but there is no sign of `GL_DEPTH_CLAMP` support so it's not useful for us.